File: test/typescript/device.test.ts
/*
 * Copyright 2021 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { test, expect } from "@jest/globals";
import { Device } from "../../dist/lib/firebolt";

test("Device.version()", () => {
  const debug: string = "Non-parsable build info for error logging only."

  const os:Device.SemanticVersion = {
    major: 0,
    minor: 1,
    patch: 0,
    readable: "Firebolt OS v0.1.0"
  }

  const sdk:Device.SemanticVersion = {
    major: 0,
    minor: 8,
    patch: 0,
    readable: "The Firebolt JS SDK",
  }

  const firmware:Device.SemanticVersion = {
    major: 1,
    minor: 2,
    patch: 3,
    readable: "Device Firmware v1.2.3"
  }

  const api:Device.SemanticVersion = {
    major: 0,
    minor: 8,
    patch: 0,
    readable: "Firebolt API v0.8.0"
  }    

  return Device.version().then((v: any) => {
    expect(v.debug).toBe(debug)
    expect(v.os).toEqual(os);
    expect(v.api).toEqual(api);
    expect(v.firmware).toEqual(firmware);
    expect(typeof v.sdk.major).toBe('number')
    expect(typeof v.sdk.minor).toBe('number')
    expect(typeof v.sdk.patch).toBe('number')
    expect(typeof v.sdk.readable).toBe('string')

  });
});

test("Device.id()", () => {
  return Device.id().then((res: string) => {
    expect(res).toBe("123");
  });
});

test("Device.distributor()", () => {
  return Device.distributor().then((res: string) => {
    expect(res).toBe("Company");
  });
});

test("Device.platform()", () => {
  return Device.platform().then((res: string) => {
    expect(res).toBe("WPE");
  });
});

test("Device.uid()", () => {
  return Device.uid().then((res: string) => {
    expect(res).toBe("ee6723b8-7ab3-462c-8d93-dbf61227998e");
  });
});

test("Device.type()", () => {
  return Device.type().then((res: string) => {
    expect(res).toBe("STB");
  });
});

test("Device.model()", () => {
  return Device.model().then((res: string) => {
    expect(res).toBe("xi6");
  });
});

test("Device.sku()", () => {
  return Device.sku().then((res: string) => {
    expect(res).toBe("AX061AEI");
  });
});

test("Device.make()", () => {
  return Device.make().then((res: string) => {
    expect(res).toBe("Arris");
  });
});

test("Device.hdcp()", () => {
  const expectedOutput: Device.BooleanMap = {
    "hdcp1.4": true,
    "hdcp2.2": true,
  };
  return Device.hdcp().then((res: Device.BooleanMap) => {
    expect(res).toEqual(expectedOutput);
  });
});

test("Device.hdcp(subscriber)", () => {
  return Device.hdcp((supportedHdrProfiles: Device.BooleanMap) => {}).then(
    (res: number) => {
      expect(res > 0).toBe(true);
    }
  );
});

test("Device.audio()", () => {
  return Device.audio().then((res: Device.AudioProfiles) => {
    expect(res).toEqual({
      dolbyAtmos: true,
      "dolbyDigital5.1": true,
      "dolbyDigital5.1+": true,
      stereo: true,
    });
  });
});

test("Device.audio(subscriber)", () => {
  return Device.audio(
    (supportedAudioProfiles: Device.AudioProfiles) => {}
  ).then((res: number) => {
    expect(res > 0).toBe(true);
  });
});

test("Device.screenResolution()", () => {
  return Device.screenResolution().then((res: [number, number]) => {
    expect(res).toEqual(expect.arrayContaining([1920, 1080]));
  });
});

test("Device.screenResolution(subscriber)", () => {
  return Device.screenResolution(
    (screenResolution: [number, number]) => {}
  ).then((res: number) => {
    expect(res > 0).toBe(true);
  });
});

test("Device.videoResolution()", () => {
  return Device.videoResolution().then((res: [number, number]) => {
    expect(res).toEqual(expect.arrayContaining([1920, 1080]));
  });
});

test("Device.videoResolution(subscriber)", () => {
  return Device.videoResolution((videoResolution: [number, number]) => {}).then(
    (res: number) => {
      expect(res > 0).toBe(true);
    }
  );
});

test("Device.name()", () => {
  return Device.name().then((res: string) => {
    expect(res).toBe("Living Room");
  });
});

test("Device.name(subscriber)", () => {
  return Device.name((value: string) => {}).then((res: number) => {
    expect(res > 0).toBe(true);
  });
});

test("Device.network()", () => {
  return Device.network().then((res: object) => {
    expect(res).toEqual({ state: "connected", type: "wifi" });
  });
});

test("Device.network(subscriber)", () => {
  return Device.network((networkInfo: object) => {}).then((res: number) => {
    expect(res > 0).toBe(true);
  });
});

test("listen() audioChanged event.", () => {
  return Device.listen("audioChanged", () => {}).then((res: number) => {
    expect(res > 0).toBe(true);
  });
});

test("once() audioChanged event.", () => {
  return Device.once("audioChanged", () => {}).then((res: number) => {
    expect(res > 0).toBe(true);
  });
});

test("listen() deviceNameChanged event.", () => {
  return Device.listen("deviceNameChanged", () => {}).then((res: number) => {
    expect(res > 0).toBe(true);
  });
});

test("once() deviceNameChanged event.", () => {
  return Device.once("deviceNameChanged", () => {}).then((res: number) => {
    expect(res > 0).toBe(true);
  });
});

test("listen() hdcpChanged event.", () => {
  return Device.listen("hdcpChanged", () => {}).then((res: number) => {
    expect(res > 0).toBe(true);
  });
});

test("once() hdcpChanged event.", () => {
  return Device.once("hdcpChanged", () => {}).then((res: number) => {
    expect(res > 0).toBe(true);
  });
});

test("listen() hdrChanged event.", () => {
  return Device.listen("hdrChanged", () => {}).then((res: number) => {
    expect(res > 0).toBe(true);
  });
});

test("once() hdrChanged event.", () => {
  return Device.once("hdrChanged", () => {}).then((res: number) => {
    expect(res > 0).toBe(true);
  });
});

test("listen() nameChanged event.", () => {
  return Device.listen("nameChanged", () => {}).then((res: number) => {
    expect(res > 0).toBe(true);
  });
});

test("once() nameChanged event.", () => {
  return Device.once("nameChanged", () => {}).then((res: number) => {
    expect(res > 0).toBe(true);
  });
});

test("listen() networkChanged event.", () => {
  return Device.listen("networkChanged", () => {}).then((res: number) => {
    expect(res > 0).toBe(true);
  });
});

test("once() networkChanged event.", () => {
  return Device.once("networkChanged", () => {}).then((res: number) => {
    expect(res > 0).toBe(true);
  });
});

test("listen() screenResolutionChanged event.", () => {
  return Device.listen("screenResolutionChanged", () => {}).then(
    (res: number) => {
      expect(res > 0).toBe(true);
    }
  );
});

test("once() screenResolutionChanged event.", () => {
  return Device.once("screenResolutionChanged", () => {}).then(
    (res: number) => {
      expect(res > 0).toBe(true);
    }
  );
});

test("listen() videoResolutionChanged event.", () => {
  return Device.listen("videoResolutionChanged", () => {}).then(
    (res: number) => {
      expect(res > 0).toBe(true);
    }
  );
});

test("once() videoResolutionChanged event.", () => {
  return Device.once("videoResolutionChanged", () => {}).then((res: number) => {
    expect(res > 0).toBe(true);
  });
});

test("clear()", () => {
  const result: boolean = Device.clear(2);
  expect(result).toBeFalsy();
});

File: src/sdks/core/src/cpp/sdk/cpptest/unit/deviceTest.cpp
/*
 * Copyright 2023 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

#include "unit.h"

class DeviceTest : public ::testing::Test
{
protected:
    JsonEngine *jsonEngine;
    Firebolt::Error error = Firebolt::Error::None;

    void SetUp() override
    {
        jsonEngine = new JsonEngine();
    }

    void TearDown() override
    {
        delete jsonEngine;
    }
};

TEST_F(DeviceTest, Id)
{

    auto actual_value = jsonEngine->get_value("Device.id");

    auto id = Firebolt::IFireboltAccessor::Instance().DeviceInterface().id(&error);

    EXPECT_EQ(error, Firebolt::Error::None) << "Failed to retrieve id from Device.id() method";
    EXPECT_EQ(id, REMOVE_QUOTES(actual_value));
}

TEST_F(DeviceTest, Distributor)
{

    auto actual_value = jsonEngine->get_value("Device.distributor");

    auto distributor = Firebolt::IFireboltAccessor::Instance().DeviceInterface().distributor(&error);

    EXPECT_EQ(error, Firebolt::Error::None) << "Failed to retrieve distributor from Device.distributor() method";
    EXPECT_EQ(distributor, REMOVE_QUOTES(actual_value));
}

TEST_F(DeviceTest, Platform)
{

    auto actual_value = jsonEngine->get_value("Device.platform");

    auto platform = Firebolt::IFireboltAccessor::Instance().DeviceInterface().platform(&error);

    EXPECT_EQ(error, Firebolt::Error::None) << "Failed to retrieve platform from Device.platform() method";
    EXPECT_EQ(platform, REMOVE_QUOTES(actual_value));
}

TEST_F(DeviceTest, Uid)
{

    auto actual_value = jsonEngine->get_value("Device.uid");

    auto uid = Firebolt::IFireboltAccessor::Instance().DeviceInterface().uid(&error);

    EXPECT_EQ(error, Firebolt::Error::None) << "Failed to retrieve uid from Device.uid() method";
    EXPECT_EQ(uid, REMOVE_QUOTES(actual_value));
}

TEST_F(DeviceTest, Type)
{

    auto actual_value = jsonEngine->get_value("Device.type");

    auto type = Firebolt::IFireboltAccessor::Instance().DeviceInterface().type(&error);

    EXPECT_EQ(error, Firebolt::Error::None) << "Failed to retrieve type from Device.type() method";
    EXPECT_EQ(type, REMOVE_QUOTES(actual_value));
}

TEST_F(DeviceTest, Model)
{

    auto actual_value = jsonEngine->get_value("Device.model");

    auto model = Firebolt::IFireboltAccessor::Instance().DeviceInterface().model(&error);

    EXPECT_EQ(error, Firebolt::Error::None) << "Failed to retrieve model from Device.model() method";
    EXPECT_EQ(model, REMOVE_QUOTES(actual_value));
}

TEST_F(DeviceTest, Sku)
{

    auto actual_value = jsonEngine->get_value("Device.sku");

    auto sku = Firebolt::IFireboltAccessor::Instance().DeviceInterface().sku(&error);

    EXPECT_EQ(error, Firebolt::Error::None) << "Failed to retrieve sku from Device.sku() method";
    EXPECT_EQ(sku, REMOVE_QUOTES(actual_value));
}

TEST_F(DeviceTest, TestDeviceMake)
{

    auto actual_value = jsonEngine->get_value("Device.make");

    auto make = Firebolt::IFireboltAccessor::Instance().DeviceInterface().make(&error);

    EXPECT_EQ(error, Firebolt::Error::None) << "Failed to retrieve make from Device.make() method";
    EXPECT_EQ(make, REMOVE_QUOTES(actual_value));
}

TEST_F(DeviceTest, Hdcp)
{
    Firebolt::Error error = Firebolt::Error::None;

    // Parsing expected JSON values into a BooleanMap
    nlohmann::json expectedJson = nlohmann::json::parse(jsonEngine->get_value("Device.hdcp"));
    Firebolt::Types::BooleanMap expectedValues;

    for (auto it = expectedJson.begin(); it != expectedJson.end(); ++it)
    {
        expectedValues[it.key()] = it.value().get<bool>();
    }

    // Getting the actual value from the DeviceInterface
    Firebolt::Device::HDCPVersionMap hdcpMap = Firebolt::IFireboltAccessor::Instance().DeviceInterface().hdcp(&error);

    EXPECT_EQ(error, Firebolt::Error::None) << "Failed to retrieve hdcp from Device.hdcp() method";

    // Convert HDCPVersionMap to BooleanMap for comparison
    Firebolt::Types::BooleanMap actualValues;
    actualValues["hdcp1.4"] = hdcpMap.hdcp1_4;
    actualValues["hdcp2.2"] = hdcpMap.hdcp2_2;

    EXPECT_EQ(actualValues, expectedValues);
}

TEST_F(DeviceTest, Hdr)
{
    Firebolt::Error error = Firebolt::Error::None;

    // Parsing expected JSON values into a BooleanMap
    nlohmann::json expectedJson = nlohmann::json::parse(jsonEngine->get_value("Device.hdr"));
    Firebolt::Types::BooleanMap expectedValues;

    for (auto it = expectedJson.begin(); it != expectedJson.end(); ++it)
    {
        expectedValues[it.key()] = it.value().get<bool>();
    }

    // Getting the actual value from the DeviceInterface
    Firebolt::Device::HDRFormatMap hdrMap = Firebolt::IFireboltAccessor::Instance().DeviceInterface().hdr(&error);

    EXPECT_EQ(error, Firebolt::Error::None) << "Failed to retrieve hdr from Device.hdr() method";

    // Convert HDRFormatMap to BooleanMap for comparison
    Firebolt::Types::BooleanMap actualValues;
    actualValues["hdr10"] = hdrMap.hdr10;
    actualValues["hdr10Plus"] = hdrMap.hdr10Plus;
    actualValues["dolbyVision"] = hdrMap.dolbyVision;
    actualValues["hlg"] = hdrMap.hlg;

    EXPECT_EQ(actualValues, expectedValues);
}

TEST_F(DeviceTest, Audio)
{ 
    nlohmann::json expectedValues = nlohmann::json::parse(jsonEngine->get_value("Device.audio"));

    // Getting the actual value from the DeviceInterface
    const Firebolt::Device::AudioProfiles audio = Firebolt::IFireboltAccessor::Instance().DeviceInterface().audio(&error);

    EXPECT_EQ(error, Firebolt::Error::None) << "Failed to retrieve audio from Device.audio() method";
    EXPECT_EQ(audio.stereo, expectedValues["stereo"]);
    EXPECT_EQ(audio.dolbyDigital5_1, expectedValues["dolbyDigital5.1"]);
    EXPECT_EQ(audio.dolbyDigital5_1_plus, expectedValues["dolbyDigital5.1+"]);
    EXPECT_EQ(audio.dolbyAtmos, expectedValues["dolbyAtmos"]);
}


std::string networkTypeToString(Firebolt::Device::NetworkType type) {
    switch (type) {
        case Firebolt::Device::NetworkType::WIFI:
            return "wifi";
        case Firebolt::Device::NetworkType::ETHERNET:
            return "ethernet";
        case Firebolt::Device::NetworkType::HYBRID:
            return "hybrid";
        default:
            return "";
    }
}

std::string networkStateToString(Firebolt::Device::NetworkState state) {
    switch (state) {
        case Firebolt::Device::NetworkState::CONNECTED:
            return "connected";
        case Firebolt::Device::NetworkState::DISCONNECTED:
            return "disconnected";
        default:
            return "";
    }
}

TEST_F(DeviceTest, Network)
{
    nlohmann::json expectedValues = nlohmann::json::parse(jsonEngine->get_value("Device.network"));

    // Getting the actual value from the DeviceInterface
    Firebolt::Device::NetworkInfoResult network = Firebolt::IFireboltAccessor::Instance().DeviceInterface().network(&error);

    // Perform the assertions
    EXPECT_EQ(error, Firebolt::Error::None) << "Failed to retrieve network from Device.network() method";
    EXPECT_EQ(networkStateToString(network.state), expectedValues["state"]);
    EXPECT_EQ(networkTypeToString(network.type), expectedValues["type"]);
}

TEST_F(DeviceTest, ScreenResolution)
{

    // Parse expected JSON values
    nlohmann::json expectedJson;
    try
    {
        std::string jsonString = jsonEngine->get_value("Device.screenResolution");
        expectedJson = nlohmann::json::parse(jsonString);
    }
    catch (const nlohmann::json::exception &e)
    {
        FAIL() << "Failed to parse JSON: " << e.what();
    }

    // Validate that expectedJson is an array and has the required number of elements
    if (!expectedJson.is_array())
    {
        FAIL() << "Expected JSON is not an array: " << expectedJson.dump(4);
    }

    // Getting the actual value from the DeviceInterface
    Firebolt::Error error = Firebolt::Error::None;
    using Resolution = std::pair<int32_t, int32_t>;
    Resolution screenResolution = Firebolt::IFireboltAccessor::Instance().DeviceInterface().screenResolution(&error);

    // Perform the assertions
    EXPECT_EQ(error, Firebolt::Error::None) << "Failed to retrieve screenResolution from Device.screenResolution() method";
    EXPECT_EQ(std::to_string(screenResolution.first) + "x" + std::to_string(screenResolution.second), expectedJson.get<std::string>());
}

TEST_F(DeviceTest, VideoResolution)
{

    // Parse expected JSON values
    nlohmann::json expectedJson;
    try
    {
        std::string jsonString = jsonEngine->get_value("Device.videoResolution");
        expectedJson = nlohmann::json::parse(jsonString);
    }
    catch (const nlohmann::json::exception &e)
    {
        FAIL() << "Failed to parse JSON: " << e.what();
    }

    // Validate that expectedJson is an array and has the required number of elements
    if (!expectedJson.is_array())
    {
        FAIL() << "Expected JSON is not an array: " << expectedJson.dump(4);
    }


    // Getting the actual value from the DeviceInterface
    Firebolt::Error error = Firebolt::Error::None;
    using Resolution = std::pair<int32_t, int32_t>;
    Resolution videoResolution = Firebolt::IFireboltAccessor::Instance().DeviceInterface().videoResolution(&error);

    // Perform the assertions
    EXPECT_EQ(error, Firebolt::Error::None) << "Failed to retrieve videoResolution from Device.videoResolution() method";
    EXPECT_EQ(std::to_string(videoResolution.first) + "x" + std::to_string(videoResolution.second), expectedJson.get<std::string>());
}

TEST_F(DeviceTest, Name)
{

    auto actual_value = jsonEngine->get_value("Device.name");

    auto name = Firebolt::IFireboltAccessor::Instance().DeviceInterface().name(&error);

    EXPECT_EQ(error, Firebolt::Error::None) << "Failed to retrieve name from Device.name() method";
    EXPECT_EQ(name, REMOVE_QUOTES(actual_value));
}
